frontend/src/main/scala/bloop/engine/tasks/CompileTask.scala

```diff
@@ -171,9 +176,72 @@ object CompileTask {
 
           // Block on the task associated with this result that sets up the read-only classes dir
           waitOnReadClassesDir.flatMap { _ =>
+            def getAllSourceInputs(project: Project): List[AbsolutePath] = {
+              import java.nio.file.Files
+              import scala.collection.JavaConverters._
+
+              val uniqueSourceDirs = project.sources
+
+              val sourceExts = Seq(".scala", ".java")
+              val unmanagedSources: mutable.Set[AbsolutePath] = mutable.Set()
+
+              uniqueSourceDirs.map(_.underlying).foreach { file =>
+                if (!Files.exists(file)) ()
+                else if (
+                  Files.isRegularFile(file) && sourceExts.exists(ext => file.toString.endsWith(ext))
+                ) {
+                  unmanagedSources.add(AbsolutePath(file))
+                } else if (Files.isDirectory(file)) {
+                  Files.walk(file).iterator().asScala.foreach { file =>
+                    if (
+                      Files.isRegularFile(file) && sourceExts
+                        .exists(ext => file.toString.endsWith(ext))
+                    ) {
+                      unmanagedSources.add(AbsolutePath(file))
+                    }
+                  }
+                }
+              }
+
+              project.sourcesGlobs.foreach { glob =>
+                Files.walk(glob.directory.underlying).iterator().asScala.foreach { file =>
+                  if (
+                    Files.isRegularFile(file) && sourceExts
+                      .exists(ext => file.toString.endsWith(ext)) && glob.matches(file)
+                  ) {
+                    unmanagedSources.add(AbsolutePath(file))
+                  }
+                }
+              }
+
+              unmanagedSources.toList
+            }
+
             // Only when the task is finished, we kickstart the compilation
-            def compile(inputs: CompileInputs) =
-              Compiler.compile(inputs, isBestEffort, isBestEffortDep)
+            def compile(inputs: CompileInputs) = {
+              val firstResult = Compiler.compile(inputs, isBestEffort, isBestEffortDep, true)
+              firstResult.flatMap {
+                case result @ Compiler.Result.Failed(
+                      _,
+                      _,
+                      _,
+                      _,
+                      Some(BestEffortProducts(_, _, recompile))
+                    ) if recompile =>
+                  // we restart the compilation, starting from scratch (without any previous artifacts)
+                  inputs.reporter.reset()
+                  val foundSrcs = getAllSourceInputs(project)
```

Comment:
This should not be needed here. We already calculate sources in other places. Maybe if we just remove analysis it will not be needed?

Reply:
Bit embarassed about this but thanks for catching. I kept looking for how to properly recompile everything, came up with this (since I saw that not everything was passed into CompileInputs.sources), and only later added the key `inputs.reporter.reset()` - which it looks like it's actually all I needed, so I was able to remove that getAllSourceInputs, thankfully.